File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/mathNet/Distributions/StudentT.java
package toolgood.algorithm.mathNet.Distributions;

import java.util.function.Function;

import toolgood.algorithm.mathNet.SpecialFunctions;
import toolgood.algorithm.mathNet.RootFinding.Brent;

public class StudentT {
    public static double CDF(double location, double scale, double freedom, double x)
    {
        //if (scale <= 0.0 || freedom <= 0.0) {
        //    throw new ArgumentException(Resources.InvalidDistributionParameters);
        //}

        // TODO JVG we can probably do a better job for Cauchy special case
        if (Double.isInfinite(freedom)) {
            return Normal.CDF(location, scale, x);
        }

        double k = (x - location) / scale;
        double h = freedom / (freedom + (k * k));
        double ib = 0.5 * SpecialFunctions.BetaRegularized(freedom / 2.0, 0.5, h);
        return x <= location ? ib : 1.0 - ib;
    }
    public static double InvCDF(double location, double scale, double freedom, double p)
    {
        //if (scale <= 0.0 || freedom <= 0.0) {
        //    throw new ArgumentException(Resources.InvalidDistributionParameters);
        //}

        // TODO JVG we can probably do a better job for Cauchy special case
        if (Double.isInfinite(freedom)) {
            return Normal.InvCDF(location, scale, p);
        }

        if (p == 0.5d) {
            return location;
        }

        // TODO PERF: We must implement this explicitly instead of solving for CDF^-1
        return Brent.FindRoot(x -> {
            double k = (x - location) / scale;
            double h = freedom / (freedom + (k * k));
            double ib = 0.5 * SpecialFunctions.BetaRegularized(freedom / 2.0, 0.5, h);
            return x <= location ? ib - p : 1.0 - ib - p;
        }, -800, 800,  1e-12);
    }

}

File: csharp/ToolGood.Algorithm/MathNet/Distributions/StudentT.cs
using ToolGood.Algorithm.MathNet.Numerics.RootFinding;

namespace ToolGood.Algorithm.MathNet.Numerics.Distributions
{
    internal sealed class StudentT
    {
        public static decimal CDF(decimal location, decimal scale, int freedom, decimal x)
        {
            var k = (x - location) / scale;
            var h = freedom / (freedom + (k * k));
            var ib = 0.5m * SpecialFunctions.BetaRegularized(freedom / 2.0m, 0.5m, h);
            return x <= location ? ib : 1.0m - ib;
        }

        public static decimal InvCDF(decimal location, decimal scale, int freedom, decimal p)
        {
            if (p == 0.5m) {
                return location;
            }

            return Brent.FindRoot(x => {
                var k = (x - location) / scale;
                var h = freedom / (freedom + (k * k));
                var ib = 0.5m * SpecialFunctions.BetaRegularized(freedom / 2.0m, 0.5m, h);
                return x <= location ? ib - p : 1.0m - ib - p;
            }, -800, 800, accuracy: 1e-12m);
        }
    }
}
